modules/simple_forms_api/app/services/simple_forms_api/form_remediation/submission_archive.rb

```diff
@@ -0,0 +1,144 @@
+        @include_manifest = config.include_manifest
+        @include_metadata = config.include_metadata
+
+        assign_defaults(options)
+        hydrate_submission_data unless submission_already_hydrated?
+
+        @form_number = JSON.parse(submission&.form_data)['form_number']
+      rescue => e
+        config.handle_error("#{self.class.name} initialization failed", e)
+      end
+
+      def build!
+        create_temp_directory!(temp_directory_path)
+        process_submission_files
+
+        return "#{submission_file_path}.pdf" if archive_type == :submission
+
+        zip_directory!(config.parent_dir, temp_directory_path)
+      rescue => e
+        config.handle_error("Failed building submission: #{id}", e)
+      end
+
+      private
+
+      attr_reader :archive_type, :attachments, :config, :file_path, :form_number, :id, :include_manifest,
+                  :include_metadata, :metadata, :submission, :temp_directory_path
+
+      def assign_defaults(options)
+        # The file paths of any hydrated attachments which were originally included in the submission
+        @attachments = options[:attachments]
+        # The local path where the submission PDF is stored
+        @file_path = options[:file_path]
+        # The FormSubmission object representing the original data payload submitted
+        @submission = options[:submission]
+        # The UUID returned from the Benefits Intake API upon original submission
+        @id = @submission&.send(config.id_type) || options[:id]
+        # Data appended to the original submission headers
+        @metadata = options[:metadata]
+        # The type of archive to be created (:submission or :remediation)
+        @archive_type = options[:type] || :remediation
+      end
+
+      def submission_already_hydrated?
+        submission && file_path && attachments && metadata
+      end
+
+      def hydrate_submission_data
+        raise "No #{config.id_type} was provided" unless id
+
+        built_submission = config.remediation_data_class.new(id:).hydrate!
+        # The local path where the submission PDF is stored
+        @file_path = built_submission.file_path
+        # The FormSubmission object representing the original data payload submitted
+        @submission = built_submission.submission
+        # The UUID returned from the Benefits Intake API upon original submission
+        @id = submission&.send(config.id_type)
+        # The file paths of any hydrated attachments which were originally included in the submission
+        @attachments = built_submission.attachments || []
+        # Data appended to the original submission headers
+        @metadata = built_submission.metadata
+      end
+
+      def process_submission_files
+        [
+          -> { write_pdf },
+          -> { write_attachments if attachments&.any? },
+          -> { write_manifest if include_manifest },
+          -> { write_metadata if include_metadata }
```

Comment:
🔪

Reply:
Just out of curiosity, is this kind of similar to using `.then` in a chain of methods?

Reply:
Yessir!